### ISSUE TITLE
fetchIdl and  `anchor idl fetch` not works the same?

### ISSUE THREAD

**Issue by tg44** [project-serum/anchor#1391]
For an example a candyMachine_v2 program randomly selected from mainnet: `CggtNXgCye2qk7fLohonNftqaKT35GkuZJwHrRghEvSF`

If you try to run: 
```
anchor idl fetch CggtNXgCye2qk7fLohonNftqaKT35GkuZJwHrRghEvSF
```
It will return a json.

But with `await Program.fetchIdl(pc, provider)` it is not:
```javascript
const candyMachineId = "CggtNXgCye2qk7fLohonNftqaKT35GkuZJwHrRghEvSF"
  const payer = anchor.web3.Keypair.generate();
  const wallet = new Wallet(payer) //why do I need a wallet anyway?
  const connection = new Connection('https://api.mainnet-beta.solana.com')
  const provider = new Provider(connection, wallet, {preflightCommitment: "recent"})
  const pc = new PublicKey(candyMachineId)
  const accountInfo = await connection.getAccountInfo(pc, 'confirmed');

  console.log(accountInfo)

  const idl = await Program.fetchIdl(pc, provider)
  console.log(idl) //here idl is null
  
  //BUT
  const idlAccount = helper.decodeIdlAccount(accountInfo.data.slice(8)); //copied from idl.js
  const inflatedIdl = inflate(idlAccount.data); //imported
  const idl2 = JSON.parse(helper.utf8Decode(inflatedIdl)); //copied from utf8.js
  console.log(idl2) //now this is a json
```

Also, I have a problem with `const state = await program.account.candyMachine.fetch(pc);` I get `Invalid account discriminator`.

Is there anything I messing up, or it is indeed a bug?

`"@project-serum/anchor": "^0.20.1",`

**Comment by acheroncrypto** [comment]
`CggtNXgCye2qk7fLohonNftqaKT35GkuZJwHrRghEvSF` is not a program account, it's a [PDA](https://docs.solana.com/terminology#program-derived-account-pda). See here on Assigned Program Id [Explorer](https://explorer.solana.com/address/CggtNXgCye2qk7fLohonNftqaKT35GkuZJwHrRghEvSF)

It works with the correct program id, the candy machine v2 program account is `cndy3Z4yapfJBmL3ShUp5exZKqR3z33thTzeNMm2gRZ`

**Comment by tg44** [comment]
Yes, I'm starting to understand these, but the main thing I wanted to point out is that the two tools work differently. For me, the cli tool returns the idl even for the PDA, and the js code is not. 

Also, it would be much more user friendly if the error message would push you to the right direction (fetchIdl could have thrown an error with "It seems to be the key you passed is not a program." or "The given key is not existing on the chain." or something like that).
